webui/react/src/services/stream/stream.ts

```diff
@@ -0,0 +1,219 @@
+      this.#pendingMsg.push(msg);
+      this.#advance();
+    };
+  }
+
+  #shouldSkip(newSub: Record<Streamable, Subscription>): boolean {
+    if (!this.#curSub) return false;
+    let same = true;
+    forEach(newSub, (val, k) => {
+      if (!this.#curSub?.[k as Streamable].spec.equals(val.spec)) same = false;
+    });
+    return same;
+  }
+
+  #sendSpec(newSub?: Record<Streamable, Subscription>): void {
+    if (!newSub) return;
+    // Skip current sub and move to next only if sub has already been sent
+    if (this.#shouldSkip(newSub) && this.#syncSent) {
+      this.#advance();
+      return;
+    }
+
+    this.#numSyncs += 1;
+    const sync_id = this.#numSyncs.toString();
+    const spec = reduce(
+      newSub,
+      (spec, ent, k) => {
+        spec['known'][k] = ent.keyCache.known();
+        spec['subscribe'][k] = {
+          ...ent.spec.toWire(),
+          since: ent.keyCache.maxSeq(),
+        };
+        return spec;
+      },
+      { known: {}, subscribe: {}, sync_id: sync_id } as Record<string, StreamContent>,
+    );
+
+    this.#curSub = newSub;
+    this.#ws && this.#ws.send(JSON.stringify(spec));
+    this.#syncSent = sync_id;
+  }
+
+  #processMsg(): void {
+    while (this.#pendingMsg.length > 0) {
+      const msg = this.#pendingMsg.shift();
+      if (!msg) break;
+      if (msg['sync_id']) {
+        if (!msg['complete']) {
+          this.#syncStarted = msg['sync_id'];
+          this.#isLoading?.(false);
+        } else {
+          this.#syncComplete = msg['sync_id'];
+          this.#isLoading?.(true);
+        }
+      } else if (this.#syncSent === this.#syncStarted) {
+        // Ignore all messages between when we send a new subscription and when the
+        // sync-start message for that subscription arrives.  These are the online
+        // updates for a subscription we no longer care about.
+
+        forEach(msg, (val, k) => {
+          if (k.includes('_deleted')) {
+            const stream_key = trimEnd(k, '_deleted') as Streamable;
+            const deleted_keys = decode_keys(val as string);
+
+            this.#curSub?.[stream_key].keyCache.delete_msg(deleted_keys);
+            this.#onDelete(stream_key, deleted_keys);
+          } else {
+            this.#curSub?.[StreamEntityMap[k]].keyCache.upsert([val.id], val.seq);
+            this.#onUpsert(msg);
+          }
+        });
+      }
+    }
+  }
+
+  #processSubscription() {
+    if (!this.#curSub && (!this.#subs || this.#subs.length === 0)) return;
+
+    let spec: Record<Streamable, Subscription> | undefined;
+    // Just connected/reconnected
+    if (!this.#syncSent) {
+      // Resend current subscription if current sync not completed
+      if (this.#curSub && (!this.#syncComplete || this.#syncStarted !== this.#syncComplete)) {
+        spec = this.#curSub;
+      } else {
+        spec = this.#subs.shift();
+      }
+      this.#sendSpec(spec);
+    } else if (this.#subs && this.#syncComplete === this.#syncSent) {
```

Comment:
how about:
- no point in a bare `this.#subs` predicate, because `this.#subs` should always be defined
- test for `this.#subs.length > 0` instead, and then use `this.#sendSpec(spec!)` and avoid the `?` in `#sendSpec(newSpec?: ...)`

Alternatively, this function could probably be clearer if you used more of the if...return I talked about in the other long comment.  It's not nearly as important here as it is there though, because the complexity of the `#advance` function is a lot higher than this one, and that one is also more fundamental to understanding the entire Stream.